### PR TITLE
[SPARK-38813][SQL][3.3] Remove TimestampNTZ type support in Spark 3.3

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Remove TimestampNTZ type support in the production code of Spark 3.3. 
To archive the goal, this PR adds the check "Utils.isTesting" in the following code branches:
- keyword "timestamp_ntz" and "timestamp_ltz" in parser
- New expressions from https://issues.apache.org/jira/browse/SPARK-35662
- Using java.time.localDateTime as the external type for TimestampNTZType
- `SQLConf.timestampType` which determines the default timestamp type of Spark SQL.

This is to minimize the code difference between the master branch. So that future users won't think TimestampNTZ is already available in Spark 3.3.
The downside is that users can still find TimestampNTZType under package `org.apache.spark.sql.types`. There should be nothing left other than this.

### Why are the changes needed?

The TimestampNTZ project is not finished yet:
* Lack Hive metastore support
* Lack JDBC support
* Need to spend time scanning the codebase to find out any missing support. The current code usages of `TimestampType` are larger than `TimestampNTZType`

### Does this PR introduce _any_ user-facing change?

No, the TimestampNTZ type is not released yet.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @beliefer @sadikovi as well.
We have great progress in the past few months. But it seems risky to release TimestampNTZ recently. I suggest postpone it to the 3.4 release.

**Comment by gengliangwang** [comment]
This one is similar with https://github.com/apache/spark/pull/33444

**Comment by cloud-fan** [comment]
I think the decision was to not support Hive tables, but JDBC is a valid concern. It's really painful to support NTZ in JDBC as we already treat the existing LTZ as NTZ in the JDBC source. We need a good migration plan.

**Review by cloud-fan — APPROVED** [review]

**Comment by beliefer** [comment]
Yes. It is very painful to supports NTZ in some data sources.

**Comment by gengliangwang** [comment]
@cloud-fan I meant Hive metastore support. I just updated the PR description.

**Comment by cloud-fan** [comment]
Spark stores table schema in table properties so Hive metastore is not an issue here. It's all about hive compatibility as we can't change the mapping between Spark timestamp and Hive timestamp, which is LTZ <-> NTZ.

**Comment by sadikovi** [comment]
Can you define the risk? Lack of support in data sources is rather a limitation that could be prioritised and resolved. Is there a risk of regression in users' queries, performance implications, or backward compatibility? 

As far as I am concerned, TimestampNTZ is not a default type anyway and you have to explicitly opt-in to use this type which is up to users. If you try writing this type in JDBC data source, I expect an error to be thrown. Is the behaviour different from that?

We can state in the release notes that support is limited to certain data sources but I would not necessarily disable it for everything.

**Comment by sadikovi** [comment]
I would suggest to "disable" the data type instead of removing it, e.g. failing queries at analysis phase if the query contains TimestampNTZ.

**Comment by cloud-fan** [comment]
@sadikovi that sounds like a simpler solution. e.g. we can do this in `CheckAnalysis` and only allow NTZ in tests.

**Comment by gengliangwang** [comment]
> I would suggest to "disable" the data type instead of removing it, e.g. failing queries at analysis phase if the query contains TimestampNTZ.
@sadikovi could you raise a PR for this? I think we need to check `create/alter table` as well.

**Comment by HyukjinKwon** [comment]
Hm, can we just simply mention that `TimestampNTZ` is unstable not, and not supported very well yet? We should remove this in PySpark API docs too at least.

**Comment by gengliangwang** [comment]
Another issue is data source support..Take Parquet as an example, Spark SQL will infer the schema as `TimestampNTZType` if the annotation `isAdjustedToUTC` is `false.
cc @cloud-fan @gatorsmile , shall we mark TimestampNTZ as unstable like @HyukjinKwon said?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -157,6 +158,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
       case _: ShowTableExtended =>
         throw QueryCompilationErrors.commandUnsupportedInV2TableError("SHOW TABLE EXTENDED")
 
+      case operator: LogicalPlan
+        if !Utils.isTesting && operator.output.exists(_.dataType.isInstanceOf[TimestampNTZType]) =>
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: I removed the `!Utils.isTesting` checked and test manually. It worked.

**Comment by gengliangwang** [comment]
As this PR is becoming big, I created two follow-ups:
* Remove PySpark support: https://issues.apache.org/jira/browse/SPARK-38828
* A new flag to control the schema inference behavior: https://issues.apache.org/jira/browse/SPARK-38829
cc @itholic @sadikovi

**Comment by itholic** [comment]
Thanks for pinging!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: I got it. LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -3191,9 +3191,10 @@ object SQLConf {
         s"and type literal. Setting the configuration as ${TimestampTypes.TIMESTAMP_NTZ} will " +
         "use TIMESTAMP WITHOUT TIME ZONE as the default type while putting it as " +
         s"${TimestampTypes.TIMESTAMP_LTZ} will use TIMESTAMP WITH LOCAL TIME ZONE. " +
-        "Before the 3.3.0 release, Spark only supports the TIMESTAMP WITH " +
+        "Before the 3.4.0 release, Spark only supports the TIMESTAMP WITH " +
         "LOCAL TIME ZONE type.")
-      .version("3.3.0")
+      .version("3.4.0")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3196]: this change should be done in the master branch as well.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3196]: Sure

**Comment by gengliangwang** [comment]
@cloud-fan @sadikovi @beliefer @HyukjinKwon Thanks for the inputs.
Merging to 3.3

**Comment by dongjoon-hyun** [comment]
Thank you for the decision!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM.

**Comment by ueshin** [comment]
A weird error message is seen as follows likely caused by this PR:

```
scala> val df = spark.range(2)
df: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> df.select("i")
org.apache.spark.sql.AnalysisException: Invalid call to dataType on unresolved object;
'Project ['i]
+- Range (0, 2, step=1, splits=Some(16))

  at org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute.dataType(unresolved.scala:137)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$4(CheckAnalysis.scala:164)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$4$adapted(CheckAnalysis.scala:164)
  at scala.collection.IndexedSeqOptimized.prefixLengthImpl(IndexedSeqOptimized.scala:41)
  at scala.collection.IndexedSeqOptimized.exists(IndexedSeqOptimized.scala:49)
  at scala.collection.IndexedSeqOptimized.exists$(IndexedSeqOptimized.scala:49)
  at scala.collection.mutable.ArrayBuffer.exists(ArrayBuffer.scala:49)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1(CheckAnalysis.scala:164)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis$1$adapted(CheckAnalysis.scala:100)
  at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:357)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.checkAnalysis(CheckAnalysis.scala:100)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.checkAnalysis$(CheckAnalysis.scala:95)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.checkAnalysis(Analyzer.scala:187)
...
```

The error message should be the following as the master branch shows:

```
org.apache.spark.sql.AnalysisException: Column 'i' does not exist. Did you mean one of the following? [id];
...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: Ah, a requirement is we should call `dataType` only on resolved attributes, so the check should be
```
operator.output.exists(attr => attr.resolved && attr.dataType.isInstanceOf[TimestampNTZType])
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: @gengliangwang can you help to fix it?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:162]: Yes I am about to create a PR

**Pull request closed without merging**